File: JavaScript/947-Most Stones Removed with Same Row or Column.js
/**
 * @param {number[][]} stones
 * @return {number}
 */
var removeStones = function(stones) {
    let parent = new Map();
    
    let find = (x) =>{
        if ( x !== parent.get(x)){
            parent.set(x, find(parent.get(x)));
        }
        return parent.get(x);
    }
    
    let union = (x, y) =>{
        if (!parent.has(x)) parent.set(x, x);
        if (!parent.has(y)) parent.set(y, y);
        
        const px = find(x), py = find(y);
        if (px !== py) parent.set(px, py);
    }
    
    for (let [x, y] of stones){
        union(x, ~y);
    }
    
    let rootSet = new Set();
    for (let p of parent.keys()){
        rootSet.add(find(p));
    }
    
    return stones.length - rootSet.size;
};

File: C++/947-Most Stones Removed with Same Row or Column.cpp
// O(n) union-find solution. Note the way we count components
class Solution {
public:
    int removeStones(vector<vector<int>>& stones) {
        DSU dsu(20002);
        for (auto &stone:stones) {
            int u = stone[0], v = stone[1];
            dsu.uni(u, v+10001);
        }

        return stones.size() - dsu.components;
    }
private:
    class DSU {
    public:
        int components;
        vector<int> parent, size;
        unordered_set<int> visited;
        DSU (int N) {
            for (int i=0; i<N; i++) parent.push_back(i);
            size.resize(N, 1);
            components = 0;
        }

        int find(int x) {
            if (!visited.contains(x)) {
                visited.insert(x);
                components++;
            }
            if (x != parent[x]) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        void uni(int a, int b) {
            int pa = find(a), pb = find(b);
            if (pa != pb) {
                if (size[pa] > size[pb]) swap(pa, pb);
                parent[pa] = pb;
                size[pb] += size[pa];
                components--;
            }
        }
    };
};